flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -3317,6 +3317,39 @@ class ArrayExprLowering {
     return genarr(asScalarRef(sym));
   }
 
+  /// Base the indices according to the origin of an entity. This is needed for
+  /// operations like fir.array_coor that do not expect zero based indices.
+  llvm::SmallVector<mlir::Value>
+  applyShift(mlir::Value shape, mlir::Value slice,
+             llvm::ArrayRef<mlir::Value> indices) {
+    if (indices.empty())
+      return {};
+    auto type = indices[0].getType();
+    auto loc = getLoc();
+    if (!shape || slice || shape.getType().isa<fir::ShapeType>()) {
```

Comment:
Why would a ShapeOp ever have a type other than `fir::ShapeType`? (Per the tablegen specification, that's not possible.)

Reply:
Shared `originateIndices` instead.